File: dynamic programming/最长上升子序列.py
# coding:utf-8
__date__ = '2018/5/16 9:33'


class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        l = len(nums)
        dp = [1 for _ in range(l)]
        for i in range(1, l):
            for j in range(0, i):
                if nums[i] > nums[j] and dp[i] < dp[j] + 1:
                    dp[i] = dp[j] + 1
        return max(dp)


s = Solution()
nums = [10, 9, 2, 5, 3, 4]
s.lengthOfLIS(nums)

File: dynamic programming/最长上升子序列.go
package main

import (
	"math"
	"fmt"
)

func LengthOfLIS(nums[] int) (res int)  {
	length := len(nums)
	dp := make([]int, length)
	// 初始化数组
	for i := 0; i < length; i++ {
		dp[i] = 1
	}

	for i := 1;i < length ; i ++  {
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] && dp[i] < dp[j] + 1 {
				dp[i] = dp[j] + 1
			}
		}
	}

	res = 0

	for _, v := range dp {
		res = int(math.Max(float64(res), float64(v)))
	}
	fmt.Println(res)
	return res

}

func main() {
	nums := [] int {1, 2, 3,5,6}
	LengthOfLIS(nums)
}